aws/s3.go

```
@@ -0,0 +1,306 @@
+
+	// Handle versioned buckets.
+	if isVersioned {
+		err := svc.ListObjectVersionsPages(
+			&s3.ListObjectVersionsInput{
+				Bucket: bucketName,
+			},
+			func(page *s3.ListObjectVersionsOutput, lastPage bool) (shouldContinue bool) {
+				for _, obj := range page.Versions {
+					logging.Logger.Debugf("Bucket %s object %s version %s", *bucketName, *obj.Key, *obj.VersionId)
+					identifiers = append(identifiers, &s3.ObjectIdentifier{
+						Key:       obj.Key,
+						VersionId: obj.VersionId,
+					})
+				}
+				for _, obj := range page.DeleteMarkers {
+					logging.Logger.Debugf("Bucket %s object %s DeleteMarker %s", *bucketName, *obj.Key, *obj.VersionId)
+					identifiers = append(identifiers, &s3.ObjectIdentifier{
+						Key:       obj.Key,
+						VersionId: obj.VersionId,
+					})
+				}
+				return true
+			},
+		)
+		return identifiers, err
+	}
+
+	// Handle non versioned buckets.
+	err := svc.ListObjectsV2Pages(
+		&s3.ListObjectsV2Input{
+			Bucket: bucketName,
+		},
+		func(page *s3.ListObjectsV2Output, lastPage bool) (shouldContinue bool) {
+			for _, obj := range page.Contents {
+				logging.Logger.Debugf("Bucket %s object %s", *bucketName, *obj.Key)
+				identifiers = append(identifiers, &s3.ObjectIdentifier{
+					Key: obj.Key,
+				})
+			}
+			return true
+		},
+	)
+
+	return identifiers, err
+}
+
+func nukeAllS3BucketObjects(svc *s3.S3, bucketName *string, batchSize int) error {
+	versioningResult, err := svc.GetBucketVersioning(&s3.GetBucketVersioningInput{
+		Bucket: bucketName,
+	})
+	if err != nil {
+		return err
+	}
+
+	isVersioned := versioningResult.Status != nil && *versioningResult.Status == "Enabled"
+
+	objects, err := getS3BucketObjects(svc, bucketName, isVersioned)
+	if err != nil {
+		return err
+	}
+
+	totalObjects := len(objects)
+
+	if totalObjects == 0 {
+		logging.Logger.Infof("Bucket: %s - empty - skipping object deletion", *bucketName)
+		return nil
+	}
+
+	if batchSize <= 0 || batchSize > 1000 {
+		batchSize = 1000
+	}
```

I think we will want to return an error instead of silently adjusting to the max.

---

Done.